### PR TITLE
deprecate `on_{train/val/test/predict}_dataloader()` from DataHooks

### LINKED ISSUE #8758: Deprecate the on_{train/val/test}_dataloader hooks
### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

The `on_train/val/test_dataloader` hooks stand out in the DataHooks protocol:
https://github.com/PyTorchLightning/pytorch-lightning/blob/963c26764682fa4cf64c93c5a7572ae0040e9c32/pytorch_lightning/core/hooks.py#L669-L670

These hooks are called immediately before we call `train/val/test/predict_dataloader` so it's extremely unclear to users why 2 sets of hooks are offered: https://github.com/PyTorchLightning/pytorch-lightning/blob/4928dc55790eec4ee4a558f494a6fc3d49a5e78a/pytorch_lightning/trainer/data_loading.py#L511-L512

@carmocca points out it's used in Flash for appending the DataPipeline, but it seems like these could be done in the regular dataloader hooks of the model.

@ethanwharris mentioned it's likely we can deprecate these hooks. @tchaton to confirm

### PR DESCRIPTION
**Pull request by ninginthecloud**
## What does this PR do?

# Context
Fixes #8758

DataHooks has defined `on_{train/val/test/predict}_dataloader` , but these functions ususally are called right before `{train/val/test/predict}_dataloader`. It's difficult for users to distinguish these two sets of hooks. As discussed in issue #8758  (detailed discussion [here](https://github.com/PyTorchLightning/pytorch-lightning/discussions/8735)), it's aligned that we can deprecate this set of hooks. 

# Steps
- remove  `on_{train/val/test/predict}_dataloader`  from DataHooks
- fix tests `test_dataloaders.py` and `test_hooks.py`
- update `fx_validator.py `
## Before submitting

- [X] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [X] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [X] Did you verify new and **existing tests pass** locally with your changes?
- [X] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/core/hooks.py [hunk 1]
```diff
@@ -680,18 +680,6 @@ def predict_dataloader(self) -> EVAL_DATALOADERS:
...
-
-    def on_val_dataloader(self) -> None:
-        """Called before requesting the val dataloader."""
-
-    def on_test_dataloader(self) -> None:
-        """Called before requesting the test dataloader."""
-
-    def on_predict_dataloader(self) -> None:
-        """Called before requesting the predict dataloader."""
-
```
Comment (ananthsub) [inline pytorch_lightning/core/hooks.py:694]: this is a removal, not a deprecation. given that these are on the stable API, we need to mark these methods as deprecated in v1.5 and that they will be removed in v1.7. it has to be a 2-stage process to allow end users a chance to migrate before the breaking change happens.

**Review by ananthsub — CHANGES_REQUESTED** [review]

File: pytorch_lightning/core/hooks.py (hunk 1 above)
Comment (ninginthecloud) [reply pytorch_lightning/core/hooks.py:694]: good point! I will keep them for now. Should I just update the docs and maybe add `rank_zero_deprecation` for the update? Similar to deprecating `running_sanity_check`?

Comment (ananthsub) [reply pytorch_lightning/core/hooks.py:694]: - Let's update the docstrings to reflect that these are deprecated
- We could add a utility function like https://github.com/PyTorchLightning/pytorch-lightning/blob/de57feff445744f431eaef29182b1bcbf58a1924/pytorch_lightning/trainer/configuration_validator.py#L22 

which accepts the lightning module and datamodule as inputs and outputs a deprecation warning if any deprecated hooks are overridden. @Borda @awaelchli what do you think?

File: pytorch_lightning/trainer/connectors/logger_connector/fx_validator.py [hunk 2]
```diff
@@ -78,11 +78,8 @@ class FxValidator:
         validation_epoch_end=dict(on_step=(False,), on_epoch=(True,)),
         test_epoch_end=dict(on_step=(False,), on_epoch=(True,)),
         configure_optimizers=None,
-        on_train_dataloader=None,
         train_dataloader=None,
-        on_val_dataloader=None,
         val_dataloader=None,
-        on_test_dataloader=None,
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/fx_validator.py:85]: why are these removed?

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -1470,35 +1471,28 @@ def __iter__(self):
...
 
         def on_train_batch_start(self, batch, batch_idx: int, dataloader_idx: int) -> None:
             assert isinstance(self.trainer.train_dataloader.loaders, DataLoaderWrapper)
             self.on_train_batch_start_called = True
 
-        def on_val_dataloader(self) -> None:
-            loader = self.val_dataloader()
-            self.val_dataloader = DataLoaderFunc(DataLoaderWrapper(loader))
-            self.on_val_dataloader_called = True
-
```
Comment (ananthsub) [inline tests/trainer/test_dataloaders.py:1501]: IMO, you could split this out to a separate PR after the deprecation. That will confirm that the initial deprecation is backwards compatible & safe, and the cleanup afterwards demonstrates how to migrate to the supported API

File: CHANGELOG.md [hunk 4]
```diff
@@ -147,6 +147,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `prepare_data_per_node` flag on Trainer and set it as a property of `DataHooks`, accessible in the `LightningModule` and `LightningDataModule` [#8958](https://github.com/PyTorchLightning/pytorch-lightning/pull/8958)
 
 
+- Deprecated `on_{train/val/test/predict}_trainer()` from `DataHooks` [#9098](https://github.com/PyTorchLightning/pytorch-lightning/pull/9098)
+
```
Comment (ananthsub) [inline CHANGELOG.md:151]: `DataHooks` is an internal class right now. Most users would not be familiar with it. We should list `LightningModule` and `LightningDataModule` explicitly

**Review by ananthsub — COMMENTED** [review]
very nice! i only have some minor comments

File: tests/models/test_hooks.py [hunk 5]
```diff
@@ -505,7 +505,6 @@ def training_step(self, batch, batch_idx):
         dict(name="Callback.on_pretrain_routine_end", args=(trainer, model)),
         dict(name="on_pretrain_routine_end"),
         dict(name="Callback.on_sanity_check_start", args=(trainer, model)),
-        dict(name="on_val_dataloader"),
```
Comment (tchaton) [inline tests/models/test_hooks.py:508]: Same here, why are they already being removed ?

File: pytorch_lightning/trainer/configuration_validator.py [hunk 6]
```diff
@@ -101,11 +120,39 @@ def __verify_eval_loop_configuration(self, model: "pl.LightningModule", stage: s
             rank_zero_warn(f"you passed in a {loader_name} but have no {step_name}. Skipping {stage} loop")
         if has_step and not has_loader:
             rank_zero_warn(f"you defined a {step_name} but have no {loader_name}. Skipping {stage} loop")
+        # ----------------------------------------------
```
Comment (awaelchli) [inline pytorch_lightning/trainer/configuration_validator.py:123]:
```suggestion

        # ----------------------------------------------
```

File: pytorch_lightning/core/hooks.py [hunk 7]
```diff
@@ -681,16 +681,52 @@ def predict_dataloader(self) -> EVAL_DATALOADERS:
...
-        """Called before requesting the val dataloader."""
+        """Called before requesting the val dataloader.
+
+        .. deprecated:: v1.5
+            :meth:`on_val_dataloader` is deprecated and will be removed in v1.7.0.
+            Please use :meth:`val_dataloader()` directly.
+        """
+        rank_zero_deprecation(
+            "Method `on_val_dataloader` in DataHooks is deprecated and will be removed in v1.7.0."
+            " Please use ``val_dataloader()`` directly. "
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:704]:
```suggestion
            " Please use `val_dataloader()` directly."
```

File: pytorch_lightning/core/hooks.py [hunk 8]
```diff
@@ -681,16 +681,52 @@ def predict_dataloader(self) -> EVAL_DATALOADERS:
...
-        """Called before requesting the test dataloader."""
+        """Called before requesting the test dataloader.
+
+        .. deprecated:: v1.5
+            :meth:`on_test_dataloader` is deprecated and will be removed in v1.7.0.
+            Please use :meth:`test_dataloader()` directly.
+        """
+        rank_zero_deprecation(
+            "Method `on_test_dataloader` in DataHooks is deprecated and will be removed in v1.7.0."
+            " Please use ``test_dataloader()`` directly. "
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:716]:
```suggestion
            " Please use `test_dataloader()` directly."
```

File: pytorch_lightning/core/hooks.py [hunk 9]
```diff
@@ -681,16 +681,52 @@ def predict_dataloader(self) -> EVAL_DATALOADERS:
...
-        """Called before requesting the predict dataloader."""
+        """Called before requesting the predict dataloader.
+
+        .. deprecated:: v1.5
+            :meth:`on_predict_dataloader` is deprecated and will be removed in v1.7.0.
+            Please use :meth:`predict_dataloader()` directly.
+        """
+        rank_zero_deprecation(
+            "Method `on_predict_dataloader` in DataHooks is deprecated and will be removed in v1.7.0."
+            " Please use ``predict_dataloader()`` directly. "
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:728]:
```suggestion
            " Please use `predict_dataloader()` directly."
```

File: pytorch_lightning/core/hooks.py [hunk 10]
```diff
@@ -681,16 +681,52 @@ def predict_dataloader(self) -> EVAL_DATALOADERS:
...
-        """Called before requesting the train dataloader."""
+        """Called before requesting the train dataloader.
+
+        .. deprecated:: v1.5
+            :meth:`on_train_dataloader` is deprecated and will be removed in v1.7.0.
+            Please use :meth:`train_dataloader()` directly.
+        """
+        rank_zero_deprecation(
+            "Method `on_train_dataloader` in DataHooks is deprecated and will be removed in v1.7.0."
+            " Please use `train_dataloader()` directly. "
```
Comment (awaelchli) [inline pytorch_lightning/core/hooks.py:692]:
```suggestion
            " Please use `train_dataloader()` directly."
```

File: tests/trainer/test_dataloaders.py [hunk 11]
```diff
@@ -1470,35 +1471,28 @@ def __iter__(self):
...
             self.on_train_batch_start_called = False
-            self.on_val_dataloader_called = False
+            self.val_dataloader_called = False
             self.on_val_batch_start_called = False
 
-        def on_train_dataloader(self) -> None:
-            loader = self.train_dataloader()
-            self.train_dataloader = DataLoaderFunc(DataLoaderWrapper(loader))
-            self.on_train_dataloader_called = True
+        def train_dataloader(self) -> TRAIN_DATALOADERS:
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:1482]: let's not test these again, this is duplicated since we already have tests that these hooks get called.
and besides, counting how many times it is called is more important to know vs. if they are called, so I would even remove the whole test here.

**Review by awaelchli — APPROVED** [review]
